src/streams.h

```diff
@@ -296,6 +296,16 @@ class CDataStream
         data.insert(data.end(), begin(), end());
         clear();
     }
+
+    /**
+     * XOR the contents of this stream with a certain key.
+     *
+     * @param[in] key    The key used to XOR the data in this stream.
+     */
+    void Xor(const std::string& key) {
```

Comment:
std::string doesn't use 'NULL terminators', they are 8-bit-clean.

Also NULL-terminated strings don't matter here at all: you should do obfuscation with a sequence of bytes.

The length of [0,0,0,0,0,0,0,0] is 8 bytes just as well as [255,255,255,255,255,255,255,255] is. If not, that's a bug. 

e.g. if your obfuscation key is [1,0,1,2,3,4,5,6], you don't want to loop just ones.